In [2]:
import pandas as pd
import numpy as np 
import pandas_datareader
import datetime as dt

# Option Simulation

## Bank of America | Long Time Value Spread

\begin{align}

V_{p,\tau} = +30C(K=32,T = 0.11) - 30C(K=32,T=0.05) - 0.35

\end{align}

Initial premium cost $\$1,050$.

In [75]:
t1 =pd.date_range(start = dt.datetime.today(),end = '2024-02-09')
S_C_T = len(t1)/365
t2 = pd.date_range(start = dt.datetime.today(),end = '2024-03-01')
L_C_T = len(t2)/365
print(f"Short call time to maturity: {S_C_T:.4f}; Long call time to maturity: {L_C_T:.4f}")

Short call time to maturity: 0.0493; Long call time to maturity: 0.1068


In [76]:
size = 30
multiplier = 100
port = np.array([[-.9,-1.02,-0.662,-.225,.02,-.027],[1.25,1.35,0.64,0.165,-0.014,.039]])*100
port = pd.DataFrame(port,columns=['Initial_Price','Cur_Price','Delta',"Gamma",'Theta','Vega'],index = ['SC','LC'])
port.loc['Port',:] = np.array(port.sum(axis = 0))*30
#port.loc['Port_Totals',:] = (multiplier*30)*(np.array(port.sum(axis = 0)))

In [77]:
port

Initial_Price  Cur_Price  Delta  Gamma  Theta  Vega
SC            -90.0     -102.0  -66.2  -22.5    2.0  -2.7
LC            125.0      135.0   64.0   16.5   -1.4   3.9
Port         1050.0      990.0  -66.0 -180.0   18.0  36.0

- In the *call calendar spread* we are initially short delta, short gamma, long theta, and long vega (<strong>time value</strong>).
- We want the short call to finish worthless at $\tau_{1}$ such that the long call still has time value left in the option. 

# Apple | Long Straddle

\begin{align}

V_{p,\tau} = +20C(K=192.5) +20P(K=192.5) - \$7.90

\end{align}

- Both options expire February 2, 2024
- Initial price for this straddle was $\$15,800$.

In [67]:
apple = np.array([[4.15,4.95,0.579,.041,-0.194,.131],[3.75,3.20,-0.424,0.041,-0.167,.131]])*100
apple_port = pd.DataFrame(apple,columns=['Initial_Price','Cur_Price','Delta',"Gamma",'Theta','Vega'],index =['C(K=192.5)','P(K=192.5)'])
apple_port.loc['Port',:] = np.array(apple_port.sum(axis =0))*20
apple_port


Initial_Price  Cur_Price  Delta  Gamma  Theta   Vega
C(K=192.5)          415.0      495.0   57.9    4.1  -19.4   13.1
P(K=192.5)          375.0      320.0  -42.4    4.1  -16.7   13.1
Port              15800.0    16300.0  310.0  164.0 -722.0  524.0

# Intel | Short Straddle

\begin{align}

V_{p,\tau} = -20C(K=48) -20P(K=48) + \$4.25

\end{align}

- Both options expire February 2, 2024

In [72]:
intel = np.array([[2.58,2.10,0.547,0.082,-.094,.033],[1.67,1.78,-.454,0.083,-.087,0.033]])*-100
intelP = pd.DataFrame(intel,columns=['Initial_Price','Cur_Price','Delta',"Gamma",'Theta','Vega'],index = ['C(48)',"P(48)"])
intelP.loc['Port',:] =np.array(intelP.sum(axis=0))*20
intelP

Initial_Price  Cur_Price  Delta  Gamma  Theta   Vega
C(48)         -258.0     -210.0  -54.7   -8.2    9.4   -3.3
P(48)         -167.0     -178.0   45.4   -8.3    8.7   -3.3
Port         -8500.0    -7760.0 -186.0 -330.0  362.0 -132.0

## How do we $\Delta$ hedge this position?

- Simply purchase $186$ shares of stock.

- Since this is at time $\tau$ and the closing price of Intel is $\$48.27$. How do we make this portfolio $\Delta=-200$ and $\Gamma=0$ nueutral using the call,put, and the asset?

In [92]:
A = np.array([[48.2,1],[8.2,8.3]])
b = np.array([0,330])
np.linalg.solve(A,b)

array([18.13164869, 21.84584105])

In [73]:
intelP['Cur_Price']-intelP['Initial_Price']

C(48)     48.0
P(48)    -11.0
Port     740.0
dtype: float64